README.md
# StripeKit
![](https://img.shields.io/badge/Swift-5.2-lightgrey.svg?style=svg)
![](https://img.shields.io/badge/SwiftNio-2-lightgrey.svg?style=svg)
![Test](https://github.com/vapor-community/stripe-kit/workflows/Test/badge.svg)

### StripeKit is a Swift package used to communicate with the [Stripe](https://stripe.com) API for Server Side Swift Apps.

## Version support
**You can check the [CHANGELOG](https://github.com/vapor-community/stripe-kit/blob/master/CHANGELOG.md) to see which version of StripeKit meets your needs.**

## Installation
To start using StripeKit, in your `Package.swift`, add the following

```swift
.package(url: "https://github.com/vapor-community/stripe-kit.git", from: "16.0.0")
```

## Using the API
Initialize the `StripeClient`

```swift
let httpClient = HTTPClient(..)
let stripe = StripeClient(httpClient: httpClient, eventLoop: eventLoop, apiKey: "sk_12345")
```

And now you have acess to the APIs via `stripe`.

The APIs you have available correspond to what's implemented.

For example to use the `charges` API, the stripeclient has a property to access that API via routes.

```swift
 stripe.charges.create(amount: 2500,
                       currency: .usd,
		               description: "A server written in swift.",
                       source: "tok_visa").flatMap { (charge) -> EventLoopFuture<Void> in
                          if charge.status == .succeeded {
                              print("New servers are on the way 🚀")
                          } else {
                              print("Sorry you have to use Node.js 🤢")
                          }
            }
```

## Expandable objects

StripeKit supports [expandable objects](https://stripe.com/docs/api/expanding_objects) via 3 property wrappers:

`@Expandable`, `@DynamicExpandable` and `@ExpandableCollection`

All API routes that can return expanded objects have an extra parameter `expand: [String]?` that allows specifying which objects to expand. 

### Usage with `@Expandable`:
1. Expanding a single field.
```swift
// Expanding a customer from creating a `PaymentIntent`.
stripeclient.paymentIntents.create(amount: 2500, currency: .usd, expand: ["customer"])
.flatMap { paymentIntent in
// Accessing the expanded `StripeCustomer` object   
 paymentIntent.$customer.email
...
}
```

2. Expanding multiple fields.
```swift
// Expanding a customer and payment method from creating a `PaymentIntent`.
stripeclient.paymentIntents.create(amount: 2500, currency: .usd, expand: ["customer", "paymentMethod"])
.flatMap { paymentIntent in
// Accessing the expanded `StripeCustomer` object   
 paymentIntent.$customer?.email // "stripe@example.com"
// Accessing the expanded `StripePaymentMethod` object
 paymentIntent.$paymentMethod?.card?.last4 // "1234"
 ...
}
```

3. Expanding nested fields.
```swift
// Expanding a payment method and its nested customer from creating a `PaymentIntent`.
stripeclient.paymentIntents.create(amount: 2500, currency: .usd, expand: ["paymentMethod.customer"])
.flatMap { paymentIntent in
// Accessing the expanded `StripePaymentMethod` object
 paymentIntent.$paymentMethod?.card?.last4 // "1234"
// Accessing the nested expanded `StripeCustomer` object   
 paymentIntent.$paymentMethod?.$customer?.email // "stripe@example.com"
 ...
}
```

4. Usage with list all. 
> Note: For list operations [expanded fields must start with `data`](https://stripe.com/docs/api/expanding_objects?lang=curl)
```swift
// Expanding a customer from listing all `PaymentIntent`s.
stripeclient.paymentIntents.listAll(filter: ["expand": ["data.customer"...]])
.flatMap { list in
 // Accessing the first `StripePaymentIntent`'s expanded `StripeCustomer` property
  list.data?.first?.$customer?.email // "stripe@example.com"
}
```

### Usage with `@DynamicExpandable`:

Some objects in stripe can be expanded into different objects.
For example:

A `StripeApplicationFee` has an `originatingTransaction` property that can be expanded into either a [charge or a transfer](https://stripe.com/docs/api/application_fees/object#application_fee_object-originating_transaction).

When expanding it you can specify which object you expect by doing the following:

```swift
stripeclient.applicationFees.retrieve(fee: "fee_1234", expand: ["originatingTransaction"])
.flatMap { applicationfee in 
    // Access the originatingTransaction as a StripeCharge
    applicationfee.$originatingTransaction(as: StripeCharge.self)?.amount // 2500
    ...
    // Access the originatingTransaction as a StripeTransfer
    applicationfee.$originatingTransaction(as: StripeTransfer.self)?.destination // acc_1234
}
```

### Usage with `@ExpandableCollection`:
1. Expanding an array of `id`s 

```swift
stripeClient.retrieve(invoice: "in_12345", expand: ["discounts"])
.flatMap { invoice in
    // Access the discounts array as `String`s
    invoice.discounts.map { print($0) } // "","","",..
    
    // Access the array of `StripeDiscount`s
    invoice.$discounts.compactMap(\.id).map { print($0) } // "di_1","di_2","di_3",...  
}
```

## Nuances with parameters and type safety
Stripe has a habit of changing APIs and having dynamic parameters for a lot of their APIs.
To accomadate for these changes, certain routes that take arguments that are `hash`s or `Dictionaries`, are represented by a Swift dictionary `[String: Any]`.

For example consider the Connect account API. 

```swift
// We define a custom dictionary to represent the paramaters stripe requires.
// This allows us to avoid having to add updates to the library when a paramater or structure changes.
let individual: [String: Any] = ["address": ["city": "New York",
					     "country": "US",
                                             "line1": "1551 Broadway",
                                             "postal_code": "10036",
	                  	             "state": "NY"],
				 "first_name": "Taylor",
			         "last_name": "Swift",
                                 "ssn_last_4": "0000",
				 "dob": ["day": "13",
					 "month": "12",
					 "year": "1989"]]
												 
let businessSettings: [String: Any] = ["payouts": ["statement_descriptor": "SWIFTFORALL"]]

let tosDictionary: [String: Any] = ["date": Int(Date().timeIntervalSince1970), "ip": "127.0.0.1"]

    stripe.connectAccounts.create(type: .custom,									
                                  country: "US",
				  email: "a@example.com",
				  businessType: .individual,
			          defaultCurrency: .usd,
				  externalAccount: "bank_token",
			          individual: individual,
				  requestedCapabilities: ["platform_payments"],
				  settings: businessSettings,
				  tosAcceptance: tosDictionary).flatMap { connectAccount in
					print("New Stripe Connect account ID: \(connectAccount.id)")			
				}
```

## Authentication via the Stripe-Account header
The first, preferred, authentication option is to use your (the platform account’s) secret key and pass a `Stripe-Account` header identifying the connected account for which the request is being made. The example request performs a refund of a  charge on behalf of a connected account:
```swift
   stripe.refunds.headers.add(name: "Stripe-Account", value: "acc_12345")
   stripe.refunds.create(charge: "ch_12345", reason: .requestedByCustomer)
```
**NOTE:** The modified headers will remain on the route instance _(refunds in this case)_ of the `StripeClient` if a reference to it is held. If you're accessing the StripeClient in the scope of a function, the headers will not be retained.

## Error Handling
None of the API calls throw errors. Instead each route returns a successful `EventLoopFuture` or a failed `EventLoopFuture`.
```swift
 stripe.charges.create(amount: 2500,
                       currency: .usd,
		       description: "A server written in swift.",
                       source: "tok_visa")
 .flatMap { (charge) -> EventLoopFuture<Void> in
	  if charge.status == .succeeded {
	      print("New servers are on the way 🚀")
	  } else {
	      print("Sorry you have to use Node.js 🤢")
	  }
  }
  .flatMapError { error in
     print("Stripe error \(error.message)")
  }
```

## Webhooks
The webhooks API is available to use in a typesafe way to pull out entities. Here's an example of listening for the payment intent webhook.
```swift
func handleStripeWebhooks(req: Request) throws -> EventLoopFuture<HTTPResponse> {

    let signature = req.headers["Stripe-Signature"]

    try StripeClient.verifySignature(payload: req.body, header: signature, secret: "whsec_1234") 
    // Stripe dates come back from the Stripe API as epoch and the StripeModels convert these into swift `Date` types.
    // Use a date and key decoding strategy to successfully parse out the `created` property and snake case strpe properties. 
    let decoder = JSONDecoder()
    decoder.dateDecodingStrategy = .secondsSince1970
    decoder.keyDecodingStrategy = .convertFromSnakeCase
    
    let event = try decoder.decode(StripeEvent.self, from: req.bodyData)
    
    switch (event.type, event.data?.object) {
    case (.paymentIntentSucceeded, .paymentIntent(let paymentIntent)):
        print("Payment capture method: \(paymentIntent.captureMethod?.rawValue)")
        return eventLoop.makeSucceededFuture(HTTPResponse(status: .ok))
        
    default: return eventLoop.makeSucceededFuture(HTTPResponse(status: .ok))
    }
}
```

## Vapor Integration
See the [Vapor helper library](https://github.com/vapor-community/stripe) to use StripeKit with Vapor.

## Whats Implemented

### Core Resources
* [x] Balance
* [x] Balance Transactions
* [x] Charges
* [x] Customers
* [x] Disputes  
* [x] Events
* [x] Files
* [x] File Links
* [x] Mandates
* [x] PaymentIntents
* [x] SetupIntents
* [x] SetupAttempts
* [x] Payouts
* [x] Refunds
* [x] Tokens
---
### Payment Methods
* [x] Payment Methods
* [x] Bank Accounts
* [x] Cards
* [x] Sources
---
### Products
* [x] Products
* [x] Prices
* [x] Coupons
* [x] Promotion Codes
* [x] Discounts
* [x] Tax Codes
* [x] Tax Rates
* [x] Shipping Rates
---
### Checkout
* [x] Sessions
---
### Billing
* [x] Credit Notes
* [x] Customer Balance Transactions
* [x] Customer Portal
* [x] Customer Tax IDs
* [x] Invoices
* [x] Invoice Items
* [x] Plans
* [x] Quotes
* [x] Quote Line Items
* [x] Subscriptions
* [x] Subscription items
* [x] Subscription Schedule
* [x] Usage Records
---
### Connect
* [x] Account
* [x] Account Links
* [x] Application Fees
* [x] Application Fee Refunds
* [x] Capabilities
* [x] Country Specs
* [x] External Accounts
* [x] Persons
* [x] Top-ups
* [x] Transfers
* [x] Transfer Reversals
---
### Fraud
* [x] Early Fraud Warnings
* [x] Reviews
* [x] Value Lists
* [x] Value List Items
---
### Issuing
* [x] Authorizations
* [x] Cardholders
* [x] Cards
* [x] Disputes
* [x] Transactions
---
### Terminal
* [x] Connection Tokens
* [x] Locations
* [x] Readers
---
### Orders
* [x] Orders
* [x] Order Items
* [x] Returns
* [x] SKUs
* [x] Ephemeral Keys
---
### Sigma
* [x] Scheduled Queries
---
### Reporting
* [x] Report Runs
* [x] Report Types
---
### Identity
* [x] VerificationSessions
* [x] VerificationReports
---
### Webhooks
* [x] Webhook Endpoints
* [x] Signature Verification

## TODO At some point
* [ ] [Idempotent Requests](https://stripe.com/docs/api/idempotent_requests)

## License
StripeKit is available under the MIT license. See the [LICENSE](LICENSE) file for more info.

## Want to help?
Feel free to submit a pull request whether it's a clean up, a new approach to handling things, adding a new part of the API, or even if it's just a typo. All help is welcomed! 😀


CHANGELOG.md
# Changelog

## 16.2.1 - 2022-5-31
* [#148](https://github.com/vapor-community/stripe-kit/pull/148)
    * Added `card_present` to payment method type.

## 16.2.0 - 2021-12-22
* [#144](https://github.com/vapor-community/stripe-kit/pull/144)
    * Fix SwiftCrypto version requirement

## 16.1.0 - 2021-12-17
* [#143](https://github.com/vapor-community/stripe-kit/pull/143) 
    * Adds support for expanding an array of models using `@ExpandableCollection`
    * Adds `checkoutSession`, `invoice`, `invoiceItem` and `promotionCode` to `StripeDiscount`.

## 16.0.0 - 2021-12-17
* [#142](https://github.com/vapor-community/stripe-kit/pull/142) ⚠️ Breaking changes ⚠️ Multiple API updates.

## 15.0.1 - 2021-11-13
* [#141](https://github.com/vapor-community/stripe-kit/pull/141) Fixes exchange rate decoding type.

## 15.0.0 - 2021-10-08
* [#140](https://github.com/vapor-community/stripe-kit/pull/140) Makes `created` on `ConnectAccount` optional.

## 14.0.0 - 2021-09-23
* [#137](https://github.com/vapor-community/stripe-kit/pull/137) 
    * Fixes expandable properties not decoding properly on swift 5.5
    * Adds better CI for Tests.
* [#138](https://github.com/vapor-community/stripe-kit/pull/138)
    * Makes StripeSessionCustomerDetails.taxIds optional.

## 13.2.0 - 2021-08-27
* [#136](https://github.com/vapor-community/stripe-kit/pull/136)
    * Added expand parameter for cloning a payment method.

## 13.1.0 - 2021-08-17
* [#134](https://github.com/vapor-community/stripe-kit/pull/134)
    * Added support for cloning a payment method.

## 13.0.0 - 2021-08-01
* [#133](https://github.com/vapor-community/stripe-kit/pull/133) 
    * ⚠️ Breaking changes ⚠️ Multiple API updates.
    * Adds support for `Quotes` and `QuoteLineItems`
    * Adds support for `VerificationSessions` and `VerificationReports`

## 12.0.2 - 2021-04-29
* [#125](https://github.com/vapor-community/stripe-kit/pull/125) Added payment method models.

* [#126](https://github.com/vapor-community/stripe-kit/pull/126) Fixed file create crash.

* [#127](https://github.com/vapor-community/stripe-kit/pull/127) Fixed expandable null value encoding.

## 12.0.0 - 2021-04-07
* [#120](https://github.com/vapor-community/stripe-kit/pull/118) ⚠️ Breaking changes ⚠️ Multiple API updates.

## 11.0.0 - 2021-02-25
* [#118](https://github.com/vapor-community/stripe-kit/pull/118) ⚠️ Breaking changes ⚠️ Multiple API updates and new support for [Customer portal configuration](https://stripe.com/docs/billing/subscriptions/integrating-customer-portal)

## 10.1.0 - 2020-11-28
* [#113](https://github.com/vapor-community/stripe-kit/pull/113) Added support for [Customer Portal](https://stripe.com/docs/api/customer_portal) API.

## 10.0.0 - 2020-11-10
* [#112](https://github.com/vapor-community/stripe-kit/pull/112) Updated to latest 2020-08-27 API version.

## 8.0.0 - 2020-07-21
* [#98](https://github.com/vapor-community/stripe-kit/pull/98) Added support for the `Price`s API.

* [#97](https://github.com/vapor-community/stripe-kit/pull/97) Updated `Subscription`s.

* [#96](https://github.com/vapor-community/stripe-kit/pull/96) Updated `Invoice`s.

* [#95](https://github.com/vapor-community/stripe-kit/pull/95) Updated `Issuing` `Transactions`.

* [#94](https://github.com/vapor-community/stripe-kit/pull/94) Updated `PaymentIntent`s.

* [#93](https://github.com/vapor-community/stripe-kit/pull/93) Added new `PaymentMethod` types and updated `Charge`s.

* [#91](https://github.com/vapor-community/stripe-kit/pull/91) Added line items for checkout sessions.

* [#90](https://github.com/vapor-community/stripe-kit/pull/90) Updated `AccountLink`s.

* [#89](https://github.com/vapor-community/stripe-kit/pull/89) Updated `ConnectAccount` and `Person` entities.

* [#88](https://github.com/vapor-community/stripe-kit/pull/88) Mark `hasMore` properties on Lists as optional.

## 7.0.2 - 2020-04-30
* [#77](https://github.com/vapor-community/stripe-kit/pull/77) Add fix for missing field on `@Expandable` properties.

## 7.0.1 - 2020-04-28
* [#75](https://github.com/vapor-community/stripe-kit/pull/75) 
    * Made payout routes public.
    * Fixed `line_items` paramater type on checkout session routes.

## 7.0.0 - 2020-04-20
* [#73](https://github.com/vapor-community/stripe-kit/pull/73) Updated Issuing API and objects to latest spec.

* [#72](https://github.com/vapor-community/stripe-kit/pull/72) Added support for [Expandable objects](https://stripe.com/docs/api/expanding_objects).


## 6.1.0 - 2020-03-26

* [#68](https://github.com/vapor-community/stripe-kit/pull/68) Fixed type mispatch for `StripeSubscriptionProrationBehavior` on subscription update route.

* [#67](https://github.com/vapor-community/stripe-kit/pull/67) Added support for webhooks signature validation.

* [#65](https://github.com/vapor-community/stripe-kit/pull/65) 
    * Made `requestedCapabilities` `Optional` on connect accounts.
    * Added `StripePaymentIntentTransferData` type for `transferData` on `StripePaymentIntent`.

## 6.0.0 - 2020-03-07
Major version release for new API version `2020-03-02`

* [#63](https://github.com/vapor-community/stripe-kit/pull/63) Added metadata support for webhook endpoints.

* [#61](https://github.com/vapor-community/stripe-kit/pull/61) Added support for `expiry_check` in Issuing Authorization.

* [#59](https://github.com/vapor-community/stripe-kit/pull/59) Added `payment_intent.processing` webhook event.

* [#57](https://github.com/vapor-community/stripe-kit/pull/57) Added `next_invoice_sequence` for `Customer`.

* [#55](https://github.com/vapor-community/stripe-kit/pull/55) Added support for listing all `Checkout Sessions`.

* [#53](https://github.com/vapor-community/stripe-kit/pull/53) Added support for `proration_behavior` on `SubscriptionSchedule`.

* [#51](https://github.com/vapor-community/stripe-kit/pull/51) Added `timezone` to `RunReportParameters`.

* [#49](https://github.com/vapor-community/stripe-kit/pull/49) Added `additional_verification` to `File`'s `purpose`.


Sources/StripeKit/Endpoints.swift
//
//  Endpoints.swift
//  Stripe
//
//  Created by Anthony Castelli on 4/13/17.
//
//

import Foundation

internal let APIBase = "https://api.stripe.com/"
internal let FilesAPIBase = "https://files.stripe.com/"
internal let APIVersion = "v1/"

internal enum StripeAPIEndpoint {
    // MARK: - TOKENS
    case tokens
    case token(String)
    
    // MARK: - REFUNDS
    case refunds
    case refund(String)
    
    // MARK: - COUPONS
    case coupons
    case coupon(String)
    
    // MARK: - PLANS
    case plan
    case plans(String)
    
    // MARK: - SOURCES
    case source
    case sources(String)
    case sourcesAttach(String)
    case sourcesDetach(String, String)
    
    // MARK: - SUBSCRIPTION ITEMS
    case subscriptionItem
    case subscriptionItems(String)
    
    // MARK: - SUBSCRIPTIONS
    case subscription
    case subscriptions(String)
    
    // MARK: - ACCOUNTS
    case account
    case accounts(String)
    case accountsReject(String)
    case accountsLoginLink(String)
        
    // MARK: - SKUS
    case sku
    case skus(String)
    
    // MARK: - PRODUCTS
    case product
    case products(String)
    
    // MARK: - ORDERS
    case order
    case orders(String)
    case ordersPay(String)
    case ordersReturn(String)
    
    // MARK: - RETURNS
    case orderReturn
    case orderReturns(String)
    
    // MARK: - INVOICES
    case invoice
    case invoices(String)
    case invoicesFinalize(String)
    case invoicesPay(String)
    case invoicesSend(String)
    case invoicesVoid(String)
    case invoicesMarkUncollectible(String)
    case invoicesLineItems(String)
    case invoicesUpcoming
    case invoicesUpcomingLineItems
    
    // MARK: - INVOICE ITEMS
    case invoiceItem
    case invoiceItems(String)
    
    // MARK: - EPHEMERAL KEYS
    case ephemeralKeys
    case ephemeralKey(String)
    
    // MARK: - TRANSFERS
    case transfer
    case transfers(String)
    case transferReversal(String)
    case transfersReversal(String,String)
    
    // MARK: - PAYOUTS
    case payout
    case payouts(String)
    case payoutsCancel(String)
        
    // MARK: - PERSONS
    case person(String)
    case persons(String, String)
    
    case applicationFee
    case applicationFees(String)
    
    case applicationFeeRefund(String)
    case applicationFeeRefunds(String, String)
    
    case externalAccount(String)
    case externalAccounts(String, String)
    
    case countrySpec
    case countrySpecs(String)
    
    case topup
    case topups(String)
    case topupsCancel(String)
    
    case review
    case reviews(String)
    case reviewsApprove(String)
    
    case valueListItem
    case valueListItems(String)
    
    case valueList
    case valueLists(String)
    
    case paymentIntents
    case paymentIntent(String)
    case paymentIntentConfirm(String)
    case paymentIntentCapture(String)
    case paymentIntentCancel(String)
    
    case paymentMethod
    case paymentMethods(String)
    case paymentMethodsAttach(String)
    case paymentMethodsDetach(String)
    
    case bankAccount(String)
    case bankAccounts(String, String)
    case bankAccountVerify(String, String)
    
    case card(String)
    case cards(String, String)
    
    case session
    case sessions(String)
    
    case discountCustomer(String)
    case discountSubscription(String)
    
    case taxid(String)
    case taxids(String, String)
    
    case taxRate
    case taxRates(String)
    
    case creditNote
    case creditNotes(String)
    case creditNotesVoid(String)
    
    case usageRecords(String)
    case usageRecordSummaries(String)
    
    case authorization
    case authorizations(String)
    case authorizationsApprove(String)
    case authorizationsDecline(String)
    
    case cardholder
    case cardholders(String)
    
    case issuingCard
    case issuingCards(String)
    case issuingCardDetails(String)
    
    case issuingDispute
    case issuingDisputes(String)
    
    case transaction
    case transactions(String)
    
    case connectionToken
    
    case location
    case locations(String)
    
    case reader
    case readers(String)
    
    case scheduledQueryRun
    case scheduledQueryRuns(String)
    
    case accountLinks
    
    var endpoint: String {
        switch self {
        case .tokens: return APIBase + APIVersion + "tokens"
        case .token(let id): return APIBase + APIVersion + "tokens/\(id)"
            
        case .refunds: return APIBase + APIVersion + "refunds"
        case .refund(let id): return APIBase + APIVersion + "refunds/\(id)"
            
        case .coupons: return APIBase + APIVersion + "coupons"
        case .coupon(let id): return APIBase + APIVersion + "coupons/\(id)"
            
        case .plan: return APIBase + APIVersion + "plans"
        case .plans(let id): return APIBase + APIVersion + "plans/\(id)"
            
        case .source: return APIBase + APIVersion + "sources"
        case .sources(let id): return APIBase + APIVersion + "sources/\(id)"
        case .sourcesAttach(let id): return APIBase + APIVersion + "customers/\(id)/sources"
        case .sourcesDetach(let customer, let source): return APIBase + APIVersion + "customers/\(customer)/sources/\(source)"
            
        case .subscriptionItem: return APIBase + APIVersion + "subscription_items"
        case .subscriptionItems(let id): return APIBase + APIVersion + "subscription_items/\(id)"
            
        case .subscription: return APIBase + APIVersion + "subscriptions"
        case .subscriptions(let id): return APIBase + APIVersion + "subscriptions/\(id)"
            
        case .account: return APIBase + APIVersion + "accounts"
        case .accounts(let id): return APIBase + APIVersion + "accounts/\(id)"
        case .accountsReject(let id): return APIBase + APIVersion + "accounts/\(id)/reject"
        case .accountsLoginLink(let id): return APIBase + APIVersion + "accounts/\(id)/login_links"
                        
        case .sku: return APIBase + APIVersion + "skus"
        case .skus(let id): return APIBase + APIVersion + "skus/\(id)"
            
        case .product: return APIBase + APIVersion + "products"
        case .products(let id): return APIBase + APIVersion + "products/\(id)"
            
        case .order: return APIBase + APIVersion + "orders"
        case .orders(let id): return APIBase + APIVersion + "orders/\(id)"
        case .ordersPay(let id): return APIBase + APIVersion + "orders/\(id)/pay"
        case .ordersReturn(let id): return APIBase + APIVersion + "orders/\(id)/returns"
            
        case .orderReturn: return APIBase + APIVersion + "order_returns"
        case .orderReturns(let id): return APIBase + APIVersion + "order_returns/\(id)"
            
        case .invoice: return APIBase + APIVersion + "invoices"
        case .invoices(let invoice): return APIBase + APIVersion + "invoices/\(invoice)"
        case .invoicesFinalize(let invoice): return APIBase + APIVersion + "invoices/\(invoice)/finalize"
        case .invoicesPay(let invoice): return APIBase + APIVersion + "invoices/\(invoice)/pay"
        case .invoicesSend(let invoice): return APIBase + APIVersion + "invoices/\(invoice)/send"
        case .invoicesVoid(let invoice): return APIBase + APIVersion + "invoices/\(invoice)/void"
        case .invoicesMarkUncollectible(let invoice): return APIBase + APIVersion + "invoices/\(invoice)/mark_uncollectible"
        case .invoicesLineItems(let invoice): return APIBase + APIVersion + "invoices/\(invoice)/lines"
        case .invoicesUpcoming: return APIBase + APIVersion + "invoices/upcoming"
        case .invoicesUpcomingLineItems: return APIBase + APIVersion + "invoices/upcoming/lines"
            
        case .invoiceItem: return APIBase + APIVersion + "invoiceitems"
        case .invoiceItems(let id): return APIBase + APIVersion + "invoiceitems/\(id)"
        
        case .ephemeralKeys: return APIBase + APIVersion + "ephemeral_keys"
        case .ephemeralKey(let id): return APIBase + APIVersion + "ephemeral_keys/\(id)"
            
        case .transfer: return APIBase + APIVersion + "transfers"
        case .transfers(let id): return APIBase + APIVersion + "transfers/\(id)"
        case .transferReversal(let id): return APIBase + APIVersion + "transfers/\(id)/reversals"
        case .transfersReversal(let transfer, let reversal): return APIBase + APIVersion + "transfers/\(transfer)/reversals/\(reversal)"
        
        case .payout: return APIBase + APIVersion + "payouts"
        case .payouts(let id): return APIBase + APIVersion + "payouts/\(id)"
        case .payoutsCancel(let id): return APIBase + APIVersion + "payouts/\(id)/cancel"
                            
        case .person(let account): return APIBase + APIVersion + "accounts/\(account)/persons"
        case .persons(let account, let person): return APIBase + APIVersion + "accounts/\(account)/persons/\(person)"
            
        case .applicationFee: return APIBase + APIVersion + "application_fees"
        case .applicationFees(let fee): return APIBase + APIVersion + "application_fees/\(fee)"
            
        case .applicationFeeRefund(let fee): return APIBase + APIVersion + "application_fees/\(fee)/refunds"
        case .applicationFeeRefunds(let fee, let refund): return APIBase + APIVersion + "application_fees/\(fee)/refunds/\(refund)"
            
        case .externalAccount(let account): return APIBase + APIVersion + "accounts/\(account)/external_accounts"
        case .externalAccounts(let account, let id): return APIBase + APIVersion + "accounts/\(account)/external_accounts/\(id)"
            
        case .countrySpec: return APIBase + APIVersion + "country_specs"
        case .countrySpecs(let country): return APIBase + APIVersion + "country_specs/\(country)"
            
        case .topup: return APIBase + APIVersion + "topups"
        case .topups(let id): return APIBase + APIVersion + "topups/\(id)"
        case .topupsCancel(let id): return APIBase + APIVersion + "topups/\(id)/cancel"
            
        case .review: return APIBase + APIVersion + "reviews"
        case .reviews(let id): return APIBase + APIVersion + "reviews/\(id)"
        case .reviewsApprove(let id): return APIBase + APIVersion + "reviews/\(id)/approve"
            
        case .valueListItem: return APIBase + APIVersion + "value_list_items"
        case .valueListItems(let id): return APIBase + APIVersion + "value_list_items/\(id)"
            
        case .valueList: return APIBase + APIVersion + "value_lists"
        case .valueLists(let id): return APIBase + APIVersion + "value_lists/\(id)"
            
        case .paymentIntents: return APIBase + APIVersion + "payment_intents"
        case .paymentIntent(let id): return APIBase + APIVersion + "payment_intents/\(id)"
        case .paymentIntentConfirm(let id): return APIBase + APIVersion + "payment_intents/\(id)/confirm"
        case .paymentIntentCapture(let id): return APIBase + APIVersion + "payment_intents/\(id)/capture"
        case .paymentIntentCancel(let id): return APIBase + APIVersion + "payment_intents/\(id)/cancel"
            
        case .paymentMethod: return APIBase + APIVersion + "payment_methods"
        case .paymentMethods(let id): return APIBase + APIVersion + "payment_method/\(id)"
        case .paymentMethodsAttach(let id): return APIBase + APIVersion + "payment_method/\(id)/attach"
        case .paymentMethodsDetach(let id): return APIBase + APIVersion + "payment_method/\(id)/detach"
            
        case .bankAccount(let customer): return APIBase + APIVersion + "customers/\(customer)/sources"
        case .bankAccounts(let customer, let bankAccount): return APIBase + APIVersion + "customers/\(customer)/sources/\(bankAccount)"
        case .bankAccountVerify(let customer, let bankAccount): return APIBase + APIVersion + "customers/\(customer)/sources/\(bankAccount)/verify"
            
        case .card(let customer): return APIBase + APIVersion + "customers/\(customer)/sources"
        case .cards(let card, let customer): return APIBase + APIVersion + "customers/\(customer)/sources/\(card)"
            
        case .session: return APIBase + APIVersion + "checkout/sessions"
        case .sessions(let session): return APIBase + APIVersion + "checkout/sessions/\(session)"
            
        case .discountCustomer(let customer): return APIBase + APIVersion + "customers/\(customer)/discount"
        case .discountSubscription(let subscription): return APIBase + APIVersion + "subscriptions/\(subscription)/discount"
            
        case .taxid(let customer): return APIBase + APIVersion + "customers/\(customer)/tax_ids"
        case .taxids(let customer, let taxid): return APIBase + APIVersion + "customers/\(customer)/tax_ids/\(taxid)"
            
        case .taxRate: return APIBase + APIVersion + "tax_rates"
        case .taxRates(let taxRate): return APIBase + APIVersion + "tax_rates/\(taxRate)"
            
        case .creditNote: return APIBase + APIVersion + "credit_notes"
        case .creditNotes(let id): return APIBase + APIVersion + "credit_notes/\(id)"
        case .creditNotesVoid(let id): return APIBase + APIVersion + "credit_notes/\(id)/void"
            
        case .usageRecords(let subscriptionItem): return APIBase + APIVersion + "subscription_items/\(subscriptionItem)/usage_records"
        case .usageRecordSummaries(let subscriptionItem): return APIBase + APIVersion + "subscription_items/\(subscriptionItem)/usage_record_summaries"
            
        case .authorization: return APIBase + APIVersion + "issuing/authorizations"
        case .authorizations(let authorization): return APIBase + APIVersion + "issuing/authorizations/\(authorization)"
        case .authorizationsApprove(let authorization): return APIBase + APIVersion + "issuing/authorizations/\(authorization)/approve"
        case .authorizationsDecline(let authorization): return APIBase + APIVersion + "issuing/authorizations/\(authorization)/decline"
            
        case .cardholder: return APIBase + APIVersion + "issuing/cardholders"
        case .cardholders(let cardholder): return APIBase + APIVersion + "issuing/cardholders/\(cardholder)"
            
        case .issuingCard: return APIBase + APIVersion + "issuing/cards"
        case .issuingCards(let card): return APIBase + APIVersion + "issuing/cards/\(card)"
        case .issuingCardDetails(let card): return APIBase + APIVersion + "issuing/cards/\(card)/details"
            
        case .issuingDispute: return APIBase + APIVersion + "issuing/disputes"
        case .issuingDisputes(let dispute): return APIBase + APIVersion + "issuing/disputes/\(dispute)"
            
        case .transaction: return APIBase + APIVersion + "issuing/transactions"
        case .transactions(let transaction): return APIBase + APIVersion + "issuing/transactions/\(transaction)"
            
        case .connectionToken: return APIBase + APIVersion + "terminal/connection_tokens"
        
        case .location: return APIBase + APIVersion + "terminal/locations"
        case .locations(let location): return APIBase + APIVersion + "terminal/locations/\(location)"
            
        case .reader: return APIBase + APIVersion + "terminal/readers"
        case .readers(let reader): return APIBase + APIVersion + "terminal/readers/\(reader)"
            
        case .scheduledQueryRun: return APIBase + APIVersion + "sigma/scheduled_query_runs"
        case .scheduledQueryRuns(let query): return APIBase + APIVersion + "sigma/scheduled_query_runs/\(query)"
        
        case .accountLinks: return APIBase + APIVersion + "account_links"
        }
    }
}
